cleanlab_studio/cli/api_service.py

```diff
@@ -187,6 +143,117 @@ async def upload_files_for_filepath_column(filepath_column: str) -> None:
         handle_api_error_from_json(json.loads(res_text))
 
 
+async def upload_files_for_filepath_column(
+    session: aiohttp.ClientSession,
+    api_key: str,
+    dataset_id: str,
+    dataset_filepath: str,
+    rows: List[List[Any]],
+    row_ids: List[Any],
+    columns: List[str],
+    filepath_column: str,
+    media_type: str,
+) -> None:
+    """
+    Uploads all files in a filepath column to S3
+
+    :param session: client session for making http requests
+    :param api_key: api key for getting presigned posts
+    :param dataset_id: id of dataset to upload files for
+    :param dataset_filepath: filepath of dataset to upload files for (can be empty string for simple image upload)
+    :param rows: dataset rows
+    :param row_ids: ids of dataset rows
+    :param columns: dataset column names
+    :param filepath_column: name of the column to upload files for
+    :param media_type: type of media to upload
+    """
+    sem = asyncio.Semaphore(MAX_PARALLEL_UPLOADS)
+    cancelled = asyncio.Event()
+    filepath_column_idx = columns.index(filepath_column)
+    filepaths = [row[filepath_column_idx] for row in rows]
+    dataset_dir: pathlib.Path = pathlib.Path(dataset_filepath).parent
+    absolute_filepaths = [
+        str(get_image_filepath(dataset_dir, row[filepath_column_idx])) for row in rows
+    ]
+    filepath_to_posts = get_presigned_posts(
+        api_key=api_key,
+        dataset_id=dataset_id,
+        filepaths=filepaths,
+        row_ids=row_ids,
+        media_type=media_type,
+    )
+    for coro in asyncio.as_completed(
+        [
+            post_file(
+                session,
+                original_filepath,
+                absolute_filepath,
+                filepath_to_posts.get(original_filepath),
+                sem,
+                cancelled,
+            )
+            for original_filepath, absolute_filepath in zip(filepaths, absolute_filepaths)
+        ]
+    ):
+        ok, original_filepath = await coro
+        if ok:
+            info(f"Uploaded {original_filepath}")
+        else:
+            # cancel tasks we don't need anymore, to give us flexibility to
+            # not abort() later but to throw an exception and keep the
+            # Python interpreter running without a bunch of leaked
+            # coroutines
+            cancelled.set()
+            abort(f"Failed to upload {original_filepath}")
+
+
+async def post_file(
+    session: aiohttp.ClientSession,
+    original_filepath: str,
+    absolute_filepath: str,
+    post_data: Any,
```

Comment:
this is a json dict right? let's make sure we're typing this correctly, instead of just using `Any`

Reply:
So it should be a JSONDict, but I'm calling this function by passing `filepath_to_posts.get(original_filepath)` which has `Optional[Any]` type, so type checking won't pass if we update this to JSONDict

Reply:
I can create a presigned_posts list outside of the list comprehension where we call `post_file` so that we can make sure the post exists for each filepath and add a type assertion before calling `post_file`

Reply:
The list creation sounds like a good idea. Remember that we don't always have to do an assertion for types (which incurs a runtime cost). We can do a cast, which is essentially free computationally at runtime. In cases where we **know** what the type is, and just need to convince mypy of it, casting is typically a better solution than a type assert